apps/emqx_ft/src/emqx_ft.erl

```diff
@@ -271,51 +222,92 @@ on_segment(#{packet_id := PacketId}, Msg, Transfer, Offset, Checksum) ->
     }),
     Segment = {Offset, Msg#message.payload},
     %% Currently synchronous.
-    %% If we want to make it async, we need to use `emqx_ft_async_reply`,
+    %% If we want to make it async, we need to use `with_responder`,
     %% like in `on_fin`.
     ?ACK_AND_PUBLISH(store_segment(Transfer, Segment)).
 
-on_fin(#{packet_id := PacketId} = TopicReplyData, Msg, Transfer, FinalSize, FinalChecksum) ->
+on_fin(
+    #{packet_id := PacketId, mode := Mode} = TopicReplyData, Msg, Transfer, FinalSize, FinalChecksum
+) ->
     ?tp(info, "file_transfer_fin", #{
         mqtt_msg => Msg,
         packet_id => PacketId,
         transfer => Transfer,
         final_size => FinalSize,
         checksum => FinalChecksum
     }),
-    %% TODO: handle checksum? Do we need it?
-    with_new_packet(
-        TopicReplyData,
-        PacketId,
-        fun() ->
-            case assemble(Transfer, FinalSize, FinalChecksum) of
-                ok ->
-                    ?ACK_AND_PUBLISH(ok);
-                %% Assembling started, packet will be acked/replied by monitor or timeout
-                {async, Pid} ->
-                    register_async_worker(Pid, TopicReplyData);
-                {error, _} = Error ->
-                    ?ACK_AND_PUBLISH(Error)
-            end
+    FinPacketKey = {self(), PacketId},
```

Comment:
With such key pattern we could publish several messages to response topic (topics?). Is this our goal, i.e. do we want to publish response once per `fin` packet, and not once per assembled transfer? Just a question, it looks like it was the same before.

Reply:
Yes, the semantic is to receive status per _command_, the published status includes packet_id as well.